File: leetcode1041/leetcode1041_my_version.java
class Solution {
    public boolean isRobotBounded(String instructions) {
        int dir = 0;
        int posX = 0, posY = 0;
        
        for (char ins: instructions.toCharArray())
        {
            if (ins == 'L')
            {
                dir--;
            }
            else if (ins == 'R')
            {
                dir++;
            }
            
            if (ins == 'G')
            {
                // go up
                if (dir % 4 == 0)
                {
                    posY++;
                }
                // go right
                else if (dir % 4 == 1 || dir % 4 == -3)
                {
                    posX++;
                }
                // go down
                else if (dir % 4 == 2 || dir % 4 == -2)
                {
                    posY--;
                }
                // go left
                else if (dir % 4 == 3 || dir % 4 == -1)
                {
                    posX--;
                }
            }
        }
        
        // after one round, still in Origin
        if (posX == 0 && posY == 0)
        {
            return true;
        }
        
        // if you have moved and only moving forward
        if (dir % 4 == 0)
        {
            return false;
        }
        else
        {
            return true;
        }
    }
}

// Runtime: 0 ms, faster than 100.00% of Java online submissions for Robot Bounded In Circle.
// Memory Usage: 37 MB, less than 70.35% of Java online submissions for Robot Bounded In Circle.
// time complexity: O(n)
// space complexity: O(1)

File: leetcode1041/leetcode1041_my_version.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Robot Bounded In Circle.
// Memory Usage: 6.1 MB, less than 77.30% of C++ online submissions for Robot Bounded In Circle.
// time complexity: O(n)
// space complexity: O(1)

class Solution
{
public:
    bool isRobotBounded(string instructions)
    {
        pair<int, int> dirs[] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int dir = 0;
        int x = 0, y = 0;

        for (char ins : instructions)
        {
            switch (ins)
            {
            case 'G':
                x += dirs[dir].first;
                y += dirs[dir].second;
                break;
            case 'L':
                dir = (dir + 1) % 4;
                break;
            case 'R':
                dir = (dir + 4 - 1) % 4;
            }
        }

        return (x == 0 && y == 0) || (dir != 0);
    }
};